### DISCUSSION TITLE
Selectively activating event handlers

### DISCUSSION THREAD

**Discussion by vpax** [zeek/zeek#1618] [Ideas]
Context: a practical problem that arises when using `-O gen-C++` to generate C++ for a set of scripts is `@if` conditional code.  Conditionals get assessed at parse time, so the script compiler doesn't get to see their presence (as it works from parsed ASTs) and therefore can't generate different code variants.  This problem has led me to analyze where in the base scripts conditionals get used, to see if they can be replaced with other coding idioms.  One common usage (though not the only one) is to conditionally specify some additional handlers for some event.

This looks something like:
> @if ( my_condition() )
> event zeek_init() { ... }
> @endif

If we had a mechanism to selectively activate event handlers, then we could replace the above idiom with something like:
> event zeek_init() &activate_on="foo" { ... }
> event zeek_pre_init()
> &nbsp; {
> &nbsp; if ( my_condition() )
> &nbsp;&nbsp; activate_handlers_bif("foo");
> &nbsp;}
Here, the semantics of the new `&activate_on="X"` attribute are to associate a specific instance (body) of an event handler with an _activation name_ (in the example, `"foo"`).  The event handler remains inactive unless a BiF (`activate_handlers_bif()`, in my example) is called with the given name.  Such names provide a way to identify specific event handler bodies for events that have multiple handlers. It also allows convenient grouping together of multiple handlers that should all be activated as a unit.

In addition, the new `zeek_pre_init()` event is raised to facilitate such activations so that they can include `zeek_init()` handlers.  Maybe instead one could just use:
> event zeek_init() &priority=9999 { if ( my_condition() ) ... }

but implementation-wise that could be a bit weird since the set of `zeek_init()` event handlers has already started executing, so it might be complicated internally to now activate a new one.

Okay, let's hear what folks think ... !

**Comment by rsmmr** [comment]
We actually used to have something very similar to this, from the CHANGES:

```
- The new attribute &group=<tag> can be associated with event handlers
  to group them together into a set that can be manipulated as a whole
  (Robin Sommer).  <tag> is a string reflecting the name given to the group.

  The built-in enable_event_group(group: string) turns on all the analyzers
  in a given group, and disable_event_group(group: string) deactivates them.
```

IIrc, this was added as part of the old load-levels works to dynamically activate analyzers. 

It got removed in 2013:

```
commit 52cd02173ddac649c7d0f3e52f3dab73f366499a
Author: Robin Sommer <robin@icir.org>
Date:   Tue Apr 9 15:54:31 2013 -0700

    Removing event groups.
````

The commit message is pretty short on context, iirc we just decided nobody is using this and removed it as part of general clean up at that time.

Looks like the `&group` attribute is still there, but unused.

I'm pretty neutral on reintroducing this, but if so, would prefer to go back to the original version. Not sure I like the pre-init event, should probably try to get the priority version to work if it doesn't anyways.

And just to say the obvious: generally, replacing `@ifdef` isn't really possible because it's main feature is that it prevents Zeek from parsing script code altogether (which could be malformed, require new features etc.)

**Reply by 0xxon** [reply]
Just to add to this - a live example where this is used for feature gating is in some widely used scripts that use constructs like this for some SSL analyzer features:

```
@if ( Version::at_least("2.6") || ( Version::number == 20500 && Version::info$commit >= 915 ) )
...
```
Stuff like that will not show up in the base scripts that we ship because there is no reason to do this; we know what Zeek version they are running on.

So - I basically just wanted to reinforce that we will need `@if` (or something similar) no matter what.

**Reply by vpax** [reply]
@rsmmr Happy to use the previous syntax.  Were event groups activated or deactivated by default?

@0xxon @rsmmr Understood re `@if` is fundamental.  I'm not looking to do away with it.  But I think it should be possible to do away with it in the base scripts, and to provide guidance to script writers regarding how to avoid it if they want their scripts to be readily compilable to C++.

**Reply by grigorescu** [reply]
> The commit message is pretty short on context, iirc we just decided nobody is using this and removed it as part of general clean up at that time.

Dug up a bit more info, which matches your recollection.

Here's the overall commit, which mentions that the event groups were broken: https://github.com/zeek/zeek/commit/e0c4bd1a82a6887b6160ab79e32a02a75d1c9119

And then there was a bit of discussion on bro-dev: https://lists.zeek.org/archives/list/zeek-dev@lists.zeek.org/thread/QS3HGCSHPL4GRVAGYXFPWQSYKQ6HPM5J/

**Reply by rsmmr** [reply]
Thanks for the context, @grigorescu. Now the challenge is to figure out *what* was actually broken about them. :-)

@vpax I'm guessing they were on by default to avoid surprises.

**Reply by awelzel** [reply]
@grigorescu , do you happen to remember the title or some key phrases from the link you mentioned a year ago?

> And then there was a bit of discussion on bro-dev: https://lists.zeek.org/archives/list/zeek-dev@lists.zeek.org/thread/QS3HGCSHPL4GRVAGYXFPWQSYKQ6HPM5J/

That link is broken and outside of the following I wasn't able to find anything relevant on mail-archive.com or on the new Zeek discourse server:
https://www.mail-archive.com/bro-dev@bro.org/msg03277.html
